substrate/frame/revive/src/lib.rs

```diff
@@ -1454,16 +1471,23 @@ where
 						result.gas_required,
 						result.storage_deposit,
 					);
+
+					match tx.clone().try_into_unsigned() {
+						Ok(tx) => unsigned_tx = Some(tx),
+						Err(_) =>
+							return Err(EthTransactError::Message("Invalid transaction".into())),
+					}
+
 					let dispatch_call: <T as Config>::RuntimeCall = crate::Call::<T>::eth_call {
 						dest,
 						value,
 						gas_limit,
 						storage_deposit_limit,
 						data: input.clone(),
-						// Since this is a dry run, we don't need to pass the signed transaction
-						// payload. Instead, use an empty vector. The signed transaction
-						// will be provided by the user when the tx is submitted.
-						payload: Vec::new(),
+						// Payload is needed to determine costs.
+						// Safe to unwrap(), it is asserted when converting to TransactionUnsigned
+						// above
+						payload: unsigned_tx.clone().unwrap().dummy_signed_payload(),
```

Comment:
not a big fan of these unwrap, can't we just compute that unsigned_payload once and not deal with optional?